### PR TITLE
Gallery preview fix and optionally skip download of urls in postprcess

### LINKED ISSUE #6247: Gallery of images doesn't render selected image
### Describe the bug

When in [Gallery](https://www.gradio.app/docs/gallery) there is >1 image, you can click on it and display a larger version. Now it does not work, it just displays an empty placeholder.

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

Demo served at https://www.gradio.app/docs/gallery or [Colab](https://colab.research.google.com/github/gradio-app/gradio/blob/main/demo/fake_gan/run.ipynb)

### Screenshot

Screenshot 2023-11-02 at 14 07 36
After selecting image
Screenshot 2023-11-02 at 14 07 42

### Logs

```shell
No logs from Gradio, just in the Chome console I get:

GET https://www.gradio.app/tmp/gradio/48ed04473ed1e1c7eb90e6317e3e3c3fba3891a7/photo-1546456073-92b9f0a8d413ixlibrb-1.2.1ixidMnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8autoformatfitcropw387q80 404 (Not Found)
```
```


### System Info

```shell
I run demos from the Gradio page or from Colab, the newest version of Gradio
Gradio version: 4.0.2
```

### Severity

Blocking usage of gradio

### LINKED ISSUE #6280: Clicking on a gallery image breaks the preview
### Describe the bug

Clicking on a gallery image breaks the preview

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

```python
import gradio as gr 

with gr.Blocks() as demo:
    cheetahs = [
        "https://upload.wikimedia.org/wikipedia/commons/0/09/TheCheethcat.jpg",
        "https://nationalzoo.si.edu/sites/default/files/animals/cheetah-003.jpg",
        "https://img.etimg.com/thumb/msid-50159822,width-650,imgsize-129520,,resizemode-4,quality-100/.jpg",
        "https://nationalzoo.si.edu/sites/default/files/animals/cheetah-002.jpg",
        "https://images.theconversation.com/files/375893/original/file-20201218-13-a8h8uq.jpg?ixlib=rb-1.1.0&rect=16%2C407%2C5515%2C2924&q=45&auto=format&w=496&fit=clip",
    ]
    gr.Gallery(value=cheetahs, columns=4)

demo.launch()
```

### Screenshot

gallery_click_preview

### Logs

_No response_

### System Info

```shell
Gradio Environment Information:
------------------------------
Operating System: Darwin
gradio version: 4.0.2
gradio_client version: 0.7.0

------------------------------------------------
gradio dependencies in your environment:

aiofiles: 23.2.1
altair: 4.2.0
fastapi: 0.101.0
ffmpy: 0.3.1
gradio-client==0.7.0 is not installed.
httpx: 0.23.0
huggingface-hub: 0.18.0
importlib-resources: 6.0.1
jinja2: 3.1.2
markupsafe: 2.1.1
matplotlib: 3.7.2
numpy: 1.26.0
orjson: 3.9.4
packaging: 22.0
pandas: 1.5.3
pillow: 9.2.0
pydantic: 2.4.0
pydub: 0.25.1
python-multipart: 0.0.6
pyyaml: 6.0
requests: 2.28.1
semantic-version: 2.10.0
tomlkit==0.12.0 is not installed.
typer: 0.9.0
typing-extensions: 4.8.0
uvicorn: 0.23.2
websockets: 11.0.3
authlib; extra == 'oauth' is not installed.
itsdangerous; extra == 'oauth' is not installed.


gradio_client dependencies in your environment:

fsspec: 2023.6.0
httpx: 0.23.0
huggingface-hub: 0.18.0
packaging: 22.0
requests: 2.28.1
typing-extensions: 4.8.0
websockets: 11.0.3
```

### Severity

Blocking usage of gradio

### LINKED ISSUE #6287: URLs not escaped during postprocess
### Describe the bug

If you return a URL with spaces from a component, postprocessing will fail

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

```python
import gradio as gr 

with gr.Blocks() as demo:
    cheetahs = [
        "https://upload.wikimedia.org/wikipedia/commons/0/09/TheCheethcat.jpg",
        "https://huggingface.co/datasets/freddyaboulton/gradio-reviews/resolve/main00015-20230906102032-7778-Wonderwoman VintageMagStyle   _lora_SDXL-VintageMagStyle-Lora_1_, Very detailed, clean, high quality, sharp image.jpg",
    ]
    gr.Gallery(value=cheetahs, columns=4)

demo.launch()
```

### Screenshot

❯ python demo/gallery_component/run.py
Traceback (most recent call last):
  File "/Users/freddy/sources/gradio/demo/gallery_component/run.py", line 12, in <module>
    gr.Gallery(value=cheetahs, columns=4)
  File "/Users/freddy/sources/gradio/gradio/component_meta.py", line 146, in wrapper
    return fn(self, **kwargs)
  File "/Users/freddy/sources/gradio/gradio/components/gallery.py", line 111, in __init__
    super().__init__(
  File "/Users/freddy/sources/gradio/gradio/component_meta.py", line 146, in wrapper
    return fn(self, **kwargs)
  File "/Users/freddy/sources/gradio/gradio/components/base.py", line 211, in __init__
    self.value = move_files_to_cache(initial_value, self)  # type: ignore
  File "/Users/freddy/sources/gradio/gradio/processing_utils.py", line 295, in move_files_to_cache
    return client_utils.traverse(data, _move_to_cache, client_utils.is_file_obj)
  File "/Users/freddy/sources/gradio/client/python/gradio_client/utils.py", line 782, in traverse
    new_obj.append(traverse(item, func, is_root))
  File "/Users/freddy/sources/gradio/client/python/gradio_client/utils.py", line 777, in traverse
    new_obj[key] = traverse(value, func, is_root)
  File "/Users/freddy/sources/gradio/client/python/gradio_client/utils.py", line 773, in traverse
    return func(json_obj)
  File "/Users/freddy/sources/gradio/gradio/processing_utils.py", line 288, in _move_to_cache
    temp_file_path = move_resource_to_block_cache(payload.path, block)
  File "/Users/freddy/sources/gradio/gradio/processing_utils.py", line 258, in move_resource_to_block_cache
    temp_file_path = save_url_to_cache(
  File "/Users/freddy/sources/gradio/gradio/processing_utils.py", line 207, in save_url_to_cache
    temp_dir = hash_url(url)
  File "/Users/freddy/sources/gradio/gradio/processing_utils.py", line 126, in hash_url
    remote = urllib.request.urlopen(url)
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/urllib/request.py", line 517, in open
    response = self._open(req, data)
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/urllib/request.py", line 534, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/urllib/request.py", line 1389, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/urllib/request.py", line 1346, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/http/client.py", line 1285, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/http/client.py", line 1296, in _send_request
    self.putrequest(method, url, **skips)
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/http/client.py", line 1130, in putrequest
    self._validate_path(url)
  File "/Users/freddy/miniconda3/envs/gradio/lib/python3.9/http/client.py", line 1230, in _validate_path
    raise InvalidURL(f"URL can't contain control characters. {url!r} "
http.client.InvalidURL: URL can't contain control characters. '/datasets/freddyaboulton/gradio-reviews/resolve/main00015-20230906102032-7778-Wonderwoman VintageMagStyle   _lora_SDXL-VintageMagStyle-Lora_1_, Very detailed, clean, high quality, sharp image.jpg' (found at least ' ')
(gradio)

### Logs

_No response_

### System Info

```shell
Gradio Environment Information:
------------------------------
Operating System: Darwin
gradio version: 4.0.2
gradio_client version: 0.7.0

------------------------------------------------
gradio dependencies in your environment:

aiofiles: 23.2.1
altair: 4.2.0
fastapi: 0.101.0
ffmpy: 0.3.1
gradio-client==0.7.0 is not installed.
httpx: 0.23.0
huggingface-hub: 0.18.0
importlib-resources: 6.0.1
jinja2: 3.1.2
markupsafe: 2.1.1
matplotlib: 3.7.2
numpy: 1.26.0
orjson: 3.9.4
packaging: 22.0
pandas: 1.5.3
pillow: 9.2.0
pydantic: 2.4.0
pydub: 0.25.1
python-multipart: 0.0.6
pyyaml: 6.0
requests: 2.28.1
semantic-version: 2.10.0
tomlkit==0.12.0 is not installed.
typer: 0.9.0
typing-extensions: 4.8.0
uvicorn: 0.23.2
websockets: 11.0.3
authlib; extra == 'oauth' is not installed.
itsdangerous; extra == 'oauth' is not installed.


gradio_client dependencies in your environment:

fsspec: 2023.6.0
httpx: 0.23.0
huggingface-hub: 0.18.0
packaging: 22.0
requests: 2.28.1
typing-extensions: 4.8.0
websockets: 11.0.3
```

### Severity

I can work around it

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Fixes #6280 , (Fixes #6247) the imgs were referencing path instead of url
Fixes #6287

Also:
- `hash_url` was downloading the entire file to compute the hash of the url. The URL is already a unique id so let's just hash that instead. This speeds up postprocessing. This is also the #6287 fix
- We should not always be downloading a url during postprocessing. Most components are set up to display URLs that are not served from the gradio server. So if postprocess returns a `FileData` with `url`, we assume its safe to serve it as-is. That's what we were doing in 3.x. If you return a `FileData` where `path` is a url, we will download it and serve it from the gradio server. This also speeds up postprocessing. If you run the `gallery_component_events` test on main, populating the gallery takes several seconds!!
- Adds an e2e test to check #6280 as well as the gallery select event

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/components/base.py [hunk 1]
```diff
@@ -167,8 +167,11 @@ def __init__(
         if not hasattr(self, "data_model"):
             self.data_model: type[GradioDataModel] | None = None
         self.temp_files: set[str] = set()
-        self.GRADIO_CACHE = os.environ.get("GRADIO_TEMP_DIR") or str(
-            Path(tempfile.gettempdir()) / "gradio"
+        self.GRADIO_CACHE = str(
+            Path(
```
Comment (freddyaboulton) [inline gradio/components/base.py:171]: Get the full resolved path for the cache

**Comment by abidlabs** [comment]
> hash_url was downloading the entire file to compute the hash of the url. The URL is already a unique id so let's just hash that instead. This speeds up postprocessing

I *think* this should be okay, although there is always a possibility that someone uploads a new file to the same URL...

File: demo/gallery_component_events/run.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import gradio as gr 
+
+with gr.Blocks() as demo:
+    cheetahs = [
+        "https://upload.wikimedia.org/wikipedia/commons/0/09/TheCheethcat.jpg",
+        "https://nationalzoo.si.edu/sites/default/files/animals/cheetah-003.jpg",
+        "https://img.etimg.com/thumb/msid-50159822,width-650,imgsize-129520,,resizemode-4,quality-100/.jpg",
+        "https://nationalzoo.si.edu/sites/default/files/animals/cheetah-002.jpg",
+        "https://images.theconversation.com/files/375893/original/file-20201218-13-a8h8uq.jpg?ixlib=rb-1.1.0&rect=16%2C407%2C5515%2C2924&q=45&auto=format&w=496&fit=clip",
```
Comment (abidlabs) [inline demo/gallery_component_events/run.py:9]: I would suggest using images from our S3 bucket so that our functional test doesn't become flaky

Comment (freddyaboulton) [reply demo/gallery_component_events/run.py:9]: yea good point!

**Comment by abidlabs** [comment]
Awesome PR @freddyaboulton! Everything looks to be fixed, thanks for adding the tests.

Re: the URL, I think its fine for now. If this becomes an issue for any users, perhaps we can expose an option to hash the url contents instead of the just the url string.

**Review by abidlabs — APPROVED** [review]

File: demo/gallery_component_events/run.py (hunk 2 above)
Comment (freddyaboulton) [reply demo/gallery_component_events/run.py]: Made the change and pushed up!

**Pull request merged**